barretenberg/cpp/src/barretenberg/commitment_schemes/ipa/ipa.hpp

```diff
@@ -630,10 +675,102 @@ template <typename Curve_> class IPA {
     static VerifierAccumulator reduce_verify_batch_opening_claim(const BatchOpeningClaim<Curve>& batch_opening_claim,
                                                                  const std::shared_ptr<VK>& vk,
                                                                  auto& transcript)
+        requires(Curve::is_stdlib_type)
     {
         const auto opening_claim = reduce_batch_opening_claim(batch_opening_claim);
         return reduce_verify_internal(vk, opening_claim, transcript);
     }
+
+    static Fr evaluate_h_poly(const std::vector<Fr>& u_chals_inv, Fr r) {
+        Fr h_eval = 1;
+        Fr r_pow = r;
+        size_t len = u_chals_inv.size();
+        for (size_t i = 0; i < len; i++) {
+            h_eval *= (Fr(1) + u_chals_inv[len - 1 - i] * r_pow);
+            r_pow *= r_pow;
+        }
+        return h_eval;
+    }
+
+    static Fr evaluate_and_accumulate_h_polys(std::vector<Fr> u_chals_inv_1, std::vector<Fr> u_chals_inv_2, Fr r, Fr alpha) {
+        auto result = evaluate_h_poly(u_chals_inv_1, r) + alpha * evaluate_h_poly(u_chals_inv_2, r);
+        return result;
+    }
+
+    static Polynomial<bb::fq> construct_poly_from_u_chals(const std::vector<bb::fq>& u_chals_inv) {
```

Comment:
a little icky that this function is using all bb::fqs, but it's meant to be called when Curve is stdlib grumpkin, but it's not meant to deal with stdlib values.